File: backend/src/models/team_model.py
from src.utils.db_utils import CreatedAt, UUIDGenerator, UpdatedAt
from src.extensions import db

class PlayerTeamModel(db.Model):
    __tablename__ = 'player_team_table'

    player_team_id = UUIDGenerator(db, 'player-team')
    player_id = db.Column(db.String, db.ForeignKey('players_table.player_id'), nullable=False)
    team_id = db.Column(db.String, db.ForeignKey('teams_table.team_id', ondelete="CASCADE"), nullable=False)
    is_ban = db.Column(db.Boolean, nullable=False, default=False)

    __table_args__ = (
        db.UniqueConstraint('player_id', 'team_id', name='unique_player_team'),
    )

    teams_where_captain = db.relationship(
        'TeamModel',
        back_populates='team_captain',
        foreign_keys='TeamModel.team_captain_id'
    )
    
    # One-to-Many
    player = db.relationship('PlayerModel', back_populates='my_teams')
    # One-to-Many
    team = db.relationship(
        'TeamModel',
        back_populates='players',
        foreign_keys=[team_id]
    )

    created_at = CreatedAt(db)
    updated_at = UpdatedAt(db)

    def to_json_for_team(self) -> dict:
        return {
            "player_team_id": self.player_team_id,
            "user_id": self.player.user_id,
            "player_id": self.player_id,
            "first_name": self.player.first_name,
            "last_name": self.player.last_name,
            "gender": self.player.gender,
            "birth_date": self.player.birth_date,
            "player_address": self.player.player_address,
            "jersey_name": self.player.jersey_name,
            "jersey_number": self.player.jersey_number,
            "position": self.player.position,
            "profile_image_url": self.player.profile_image_url,
            "is_ban": self.is_ban
        }

    def to_json(self) -> dict:
        return {
            "player_team_id": self.player_team_id,
            "player_id": self.player_id,
            "team_id": self.team_id,
            "player": self.player.to_json() if self.player else None,
            "team": self.team.to_json() if self.team else None,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat()
        }


class TeamModel(db.Model):
    __tablename__ = 'teams_table'

    team_id = UUIDGenerator(db, 'team')

    user_id = db.Column(
        db.String,
        db.ForeignKey('users_table.user_id', ondelete='CASCADE'),
        nullable=False
    )
    # Many-to-One
    user = db.relationship('UserModel', back_populates='teams', single_parent=True)

    active_league_id = db.Column(db.String, db.ForeignKey('leagues_table.league_id'))
    active_league = db.relationship('LeagueModel', uselist=False)

    team_name = db.Column(db.String(100), nullable=False)

    team_address = db.Column(
        db.String(250),
        nullable=False
    )

    contact_number = db.Column(
        db.String(15),
        nullable=False,
    )
    
    team_motto = db.Column(
        db.String(100),
        nullable=True
    )
    
    team_logo_url = db.Column(db.String, nullable=False)
    championships_won = db.Column(db.Integer, nullable=False, default=0)
    coach_name = db.Column(db.String(100), nullable=False)
    assistant_coach_name = db.Column(db.String(100), nullable=True)

    total_wins = db.Column(db.Integer, default=0, nullable=False)
    total_losses = db.Column(db.Integer, default=0, nullable=False)

    team_captain_id = db.Column(
        db.String,
        db.ForeignKey('player_team_table.player_team_id', use_alter=True, name='fk_team_captain_id', deferrable=True),
        nullable=True
    )

    team_captain = db.relationship(
        'PlayerTeamModel',
        uselist=False,
        primaryjoin="TeamModel.team_captain_id == PlayerTeamModel.player_team_id"
    )
    
    def players_to_json_list(self):
        return [player.to_json_for_team() for player in self.players] if self.players else []

    def to_json(self):
        players = [player.to_json_for_team() for player in self.players] if self.players else []
        team_captain = self.team_captain.to_json_for_team() if self.team_captain else None
        return {
            "team_id": self.team_id,
            "user_id": self.user_id,
            "active_league": self.active_league.to_json() if self.active_league else None,
            "team_name": self.team_name,
            "team_address": self.team_address,
            "contact_number": self.contact_number,
            "team_motto": self.team_motto if self.team_motto else None,
            "team_logo_url": self.team_logo_url if self.team_logo_url else None,
            "championships_won": self.championships_won,
            "coach_name": self.coach_name,
            "assistant_coach_name": self.assistant_coach_name if self.assistant_coach_name else None,
            "team_captain": team_captain,
            "players": players,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }
    
    # Many-to-Many
    players = db.relationship(
        'PlayerTeamModel',
        back_populates='team',
        cascade='all, delete-orphan',
        foreign_keys=[PlayerTeamModel.team_id],
        passive_deletes=True
    )

    payments = db.relationship('PaymentModel', back_populates='team', cascade="all, delete-orphan")

    league_team = db.relationship(
        'LeagueTeamModel',
        back_populates='team',
        cascade='all, delete-orphan'
    )

    # use for founded_at
    created_at = CreatedAt(db)
    updated_at = UpdatedAt(db)

    

File: bogoballers/lib/core/models/team_model.dart
// ignore_for_file: non_constant_identifier_names
import 'package:bogoballers/core/models/league_model.dart';
import 'package:bogoballers/core/models/player_model.dart';
import 'package:dio/dio.dart';

class PlayerTeamModel {
  late String player_team_id;
  String player_id;
  String team_id;
  late bool is_ban;
  late PlayerModel player;
  late DateTime created_at;
  late DateTime updated_at;

  PlayerTeamModel({
    required this.player_team_id,
    required this.player_id,
    required this.team_id,
    required this.is_ban,
    required this.player,
    required this.created_at,
    required this.updated_at,
  });

  PlayerTeamModel.create({required this.player_id, required this.team_id});

  factory PlayerTeamModel.fromJson(Map<String, dynamic> json) {
    return PlayerTeamModel(
      player_team_id: json['player_team_id'],
      player_id: json['player_id'],
      team_id: json['team_id'],
      is_ban: json['is_ban'],
      player: json['player'],
      created_at: DateTime.parse(json['created_at']),
      updated_at: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'player_team_id': player_team_id,
      'player_id': player_id,
      'team_id': team_id,
      'is_ban': is_ban,
      'player': player.toJson(),
      'created_at': created_at,
      'updated_at': updated_at,
    };
  }
}

class TeamModel {
  late String team_id;
  String user_id;
  String team_name;
  String team_address;
  String contact_number;
  String team_motto;
  late String team_logo_url;
  late int championships_won;
  String coach_name;
  String? assistant_coach_name;
  late int total_wins;
  late int total_losses;
  late bool is_recruiting;
  String? team_captain_id;
  PlayerTeamModel? team_captain;
  LeagueModel? active_league;
  List<PlayerModel> players;

  late MultipartFile team_logo_image;
  late DateTime created_at;
  late DateTime updated_at;

  factory TeamModel.fromJson(Map<String, dynamic> json) {
    return TeamModel(
      team_id: json['team_id'],
      user_id: json['user_id'],
      active_league: json['active_league'] ?? null,
      team_name: json['team_name'],
      team_address: json['team_address'],
      contact_number: json['contact_number'],
      team_motto: json['team_motto'],
      team_logo_url: json['team_logo_url'],
      championships_won: json['championships_won'],
      coach_name: json['coach_name'],
      assistant_coach_name: json['assistant_coach_name'] ?? null,
      team_captain: json['team_captain'] ?? null,
      total_wins: json['total_wins'],
      total_losses: json['total_losses'] ?? 0,
      is_recruiting: json['is_recruiting'] ?? 0,
      created_at: DateTime.parse(json['created_at']),
      updated_at: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'team_id': team_id,
      'user_id': user_id,
      'active_league': active_league.toString(),
      'team_name': team_name,
      'team_address': team_address,
      'contact_number': contact_number,
      'team_motto': team_motto,
      'team_logo_url': team_logo_url,
      'championships_won': championships_won,
      'coach_name': coach_name,
      'assistant_coach_name': assistant_coach_name,
      'team_captain': team_captain?.toMap() ?? null,
      'players': players.toList(),
      'total_wins': total_wins,
      'is_recruiting': is_recruiting,
      'created_at': created_at,
      'updated_at': updated_at,
    };
  }

  TeamModel({
    required this.team_id,
    required this.user_id,
    required this.team_name,
    required this.team_address,
    required this.contact_number,
    required this.team_motto,
    required this.coach_name,
    this.active_league,
    this.assistant_coach_name,
    required this.team_logo_url,
    required this.championships_won,
    this.team_captain_id,
    required this.created_at,
    required this.updated_at,
    required this.is_recruiting,
    this.team_captain,
    this.players = const [],
    required this.total_wins,
    required this.total_losses,
  });

  TeamModel.create({
    required this.user_id,
    required this.team_name,
    required this.team_address,
    required this.contact_number,
    required this.team_motto,
    required this.coach_name,
    this.assistant_coach_name,
    required this.team_logo_image,
    this.players = const [],
  });

  get imageUrl => null;

  FormData toFormDataForCreation() {
    final formMap = {
      'user_id': user_id,
      'team_name': team_name,
      'team_address': team_address,
      'contact_number': contact_number,
      'team_motto': team_motto,
      'coach_name': coach_name,
      'assistant_coach_name': assistant_coach_name,
      'team_logo_image': team_logo_image,
    };

    return FormData.fromMap(formMap);
  }
}
